packages/compass-crud/src/components/json-editor.tsx

```diff
@@ -76,10 +76,22 @@ const JSONEditor: React.FunctionComponent<JSONEditorProps> = ({
   const [expanded, setExpanded] = useState<boolean>(doc.expanded);
   const [editing, setEditing] = useState<boolean>(doc.editing);
   const [deleting, setDeleting] = useState<boolean>(doc.markedForDeletion);
-  const [value, setValue] = useState<string>(() => doc.toEJSON());
+  const [value, setValue] = useState<string>(
+    () => doc.modifiedEJSONString ?? doc.toEJSON()
+  );
   const [initialValue] = useState<string>(() => doc.toEJSON());
   const [containsErrors, setContainsErrors] = useState<boolean>(false);
 
+  useEffect(() => {
+    return () => {
+      // When this component is used in virtualized list, the editor is
+      // unmounted on scroll and if the user is editing the document, the
+      // editor value is lost. This is a way to keep track of the editor
+      // value when the it's unmounted and is restored on next mount.
+      doc.setModifiedEJSONString(editing ? value : null);
+    };
+  }, [value, editing, doc]);
```

Comment:
maybe a docRef would make sense here? since we don't need to react to doc changes here, only value and editing